flepimop/gempyor_pkg/src/gempyor/distributions.py

```diff
@@ -0,0 +1,434 @@
+               [1.83818339, 1.9153892 , 0.37639339, 1.09435167, 0.92629918],
+               [2.54134935, 1.52545861, 2.04022114, 1.07959285, 0.6142512 ]])
+    """
+    # pylint: enable=line-too-long
+
+    distribution: Literal["truncnorm"] = "truncnorm"
+    mean: float
+    sd: float = Field(..., gt=0)
+    a: float
+    b: float
+
+    def _sample_from_generator(
+        self, size: int | tuple[int, ...], rng: Generator
+    ) -> npt.NDArray[np.float64]:
+        """Sampling logic for truncated normal distributions."""
+        if (
+            isclose(self.a, self.b) and self.allow_edge_cases
+        ):  # use this logic b/c scipy.truncnorm doesn't support equal bounds
+            return np.full(size, self.a)
+
+        lower = (self.a - self.mean) / self.sd
+        upper = (self.b - self.mean) / self.sd
+        return truncnorm.rvs(
+            a=lower,
+            b=upper,
+            loc=self.mean,
+            scale=self.sd,
+            size=size,
+            random_state=rng,
+        )
+
+    @model_validator(mode="after")
+    def _validate_bounds(self) -> "TruncatedNormalDistribution":
+        """Validate bounds based on whether or not edge cases are allowed.."""
+        if self.allow_edge_cases:
+            if self.b < self.a:
+                raise ValueError(
+                    f"Upper bound 'b' ({self.b}) must be ≥ to lower bound 'a' ({self.a})."
+                )
+        else:
+            if self.b < self.a or isclose(self.b, self.a):
+                raise ValueError(
+                    f"Upper bound 'b' ({self.b}) must be > lower bound 'a' ({self.a})."
+                )
+        return self
+
+
+class PoissonDistribution(DistributionABC):
+    """
+    Represents a Poisson distribution.
+
+    Examples:
+        >>> import numpy as np
+        >>> from gempyor.distributions import PoissonDistribution
+        >>> rng = np.random.default_rng(42)
+        >>> dist = PoissonDistribution(lam=3.0)
+        >>> dist
+        PoissonDistribution(distribution='poisson', lam=3.0)
+        >>> dist.sample(rng=rng)
+        array([4])
+        >>> dist.sample(size=(3, 5), rng=rng)
+        array([[4, 5, 1, 7, 1],
+               [4, 2, 2, 5, 4],
+               [1, 6, 2, 5, 0]])
+    """
+
+    distribution: Literal["poisson"] = "poisson"
+    lam: float
+
+    def _sample_from_generator(
+        self, size: int | tuple[int, ...], rng: Generator
+    ) -> npt.NDArray[np.int64]:
+        """Sampling logic for Poisson distributions."""
+        return rng.poisson(lam=self.lam, size=size)
+
+    @model_validator(mode="after")
+    def _validate_lambda(self) -> "PoissonDistribution":
+        if self.allow_edge_cases:
+            if self.lam < 0:
+                raise ValueError("Input for 'lam' must be ≥ 0.")
+        else:
+            if self.lam <= 0:
+                raise ValueError("Input for 'lam' must be > 0.")
+        return self
```

Comment:
```suggestion
    distribution: Literal["poisson"] = "poisson"
    lam: float = Field(ge=0.0)

    def _sample_from_generator(
        self, size: int | tuple[int, ...], rng: Generator
    ) -> npt.NDArray[np.int64]:
        """Sampling logic for Poisson distributions."""
        return rng.poisson(lam=self.lam, size=size)

    @model_validator(mode="after")
    def _validate_lambda(self) -> "PoissonDistribution":
        if not self.allow_edge_cases and isclose(self.lam, 0.0):
          raise ValueError("Input for `lam` cannot be zero when `allow_edge_cases` is `False`.")
        return self
```
Lean on pydantic more, limit custom validation